router/core/transport.go

```diff
@@ -84,68 +96,159 @@ func (ct *CustomTransport) RoundTrip(req *http.Request) (*http.Response, error)
 	return resp, err
 }
 
+func isRunningInsideDocker() bool {
+	// Check if we are running inside docker by
+	// testing by checking if /.dockerenv exists
+	st, err := os.Stat("/.dockerenv")
+	return err == nil && !st.IsDir()
+}
+
+// localhostFallbackRoundTripper is an http.RoundTripper that will retry failed
+// requests to localhost by rewriting the request to use is targetHost. Only
+// requests that fail with ECONNREFUSED will be retried.
+type localhostFallbackRoundTripper struct {
+	targetHost string
+	transport  http.RoundTripper
+}
+
+func (*localhostFallbackRoundTripper) pointsToLocalhost(r *http.Request) bool {
+	host, _, err := net.SplitHostPort(r.Host)
+	if err != nil {
+		host = r.Host
+	}
+	ip := net.ParseIP(host)
+	if ip != nil {
+		return ip.IsLoopback()
+	}
+	return host == "localhost"
+}
+
+func (t *localhostFallbackRoundTripper) rewriteToTargetHost(r *http.Request) (*http.Request, error) {
+	var newHost string
+	_, port, err := net.SplitHostPort(r.Host)
+	if err == nil {
+		newHost = t.targetHost + ":" + port
+	} else {
+		newHost = t.targetHost
+	}
+	newReq, err := http.NewRequestWithContext(r.Context(), r.Method, fmt.Sprintf("%s://%s%s", r.URL.Scheme, newHost, r.URL.Path), r.Body)
+	if err != nil {
+		return nil, err
+	}
+	newReq.Header = r.Header
+	return newReq, nil
+}
+
+func (t *localhostFallbackRoundTripper) RoundTrip(r *http.Request) (*http.Response, error) {
+	// If the request has a body, we need to buffer it, otherwise it will
+	// get consumed
+	var bodyData []byte
+	if r.Body != nil {
+		var err error
+		bodyData, err = io.ReadAll(r.Body)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read request body: %w", err)
+		}
+		r.Body.Close()
```

Comment:
For reference: Very unlikely but there are solutions like docker swarm or similar that use docker for production workloads. We should ensure to not create a solution that might conflict with it.